src/librustc_privacy/lib.rs

```diff
@@ -733,6 +675,33 @@ impl<'a, 'tcx> Visitor<'tcx> for TypePrivacyVisitor<'a, 'tcx> {
         self.tables = orig_tables;
     }
 
+    fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty) {
+        self.span = hir_ty.span;
+        if let Some(ty) = self.tables.node_id_to_type_opt(hir_ty.hir_id) {
+            // Types in bodies.
+            if ty.visit_with(self) {
+                return;
+            }
+        } else {
+            // Types in signatures.
+            if rustc_typeck::hir_ty_to_ty(self.tcx, hir_ty).visit_with(self) {
```

Comment:
This is very ineffective.
What I would actually like is to resurrect global [ast_ty_to_ty_cache](https://github.com/rust-lang/rust/pull/41733/commits/3f2dd4d24a1fae4a985ab360028b42ca1e9c61e9), but I don't know how to do it in the new query-oriented world.

Reply:
It wouldn't be global, it would be per-item and it wouldn't be a cache but an eagerly computed map.

Reply:
eagerly computed? do you mean lazilly computed?

Reply:
so .. why *can't* we put this information into the table? just because the type checker doesn't happen to include it now? It seems like we could begin the type-checking process by invoking `hir_ty_to_ty` on the argument types in the same way we do here. Not sure if that would be *better*.

Reply:
I mean eagerly computed per-item, because you have to return the entire map.

Reply:
ah, ok, yes.